index.d.ts

```diff
@@ -831,40 +831,40 @@ declare module "fastest-validator" {
 		};
 	}
 
-	interface CheckerFunctionError {
+	export interface CheckerFunctionError {
 		type: string;
 		expected?: unknown;
 		actual?: unknown;
 		field?: string;
 	}
 
-	type CheckerFunctionV1<T = unknown> = (
+	export type CheckerFunctionV1<T = unknown> = (
 		value: T,
-		schema: ValidationSchema,
+		ruleSchema: ValidationRule,
 		path: string,
 		parent: object | null,
 		context: Context
 	) => true | ValidationError[];
-	type CheckerFunctionV2<T = unknown> = (
+	export type CheckerFunctionV2<T = unknown> = (
 		value: T,
 		errors: CheckerFunctionError[],
-		schema: ValidationSchema,
+		ruleSchema: ValidationRule,
```

Comment:
better to be:
```ts
ruleSchema: ValidationRule | any
```
Because in can cause problem, if it's a custom rule, with custom keys

Reply:
Since `ValidationRule` is a union containing `ValidationRuleObject` which is a union containing `RuleCustom` and the custom rule already has an index signature, do you think that change is needed?

Is `ValidationRule` correct for the rule schema signature?  That is, can it be any of those union types, or will it be better defined than that?

Reply:
look at this sample:
```ts
const custom: CheckerFunctionV2 = function(value , errors , schema) {
   console.log("custom -> schema.foo", schema.foo); // Error
   return value;
}

const check = v.compile({
   a: {
      type: "custom",
      foo: "bar",
      custom,
   }
});
```
I got this TypeError: `Property 'foo' does not exist on type 'ValidationRule'`

Reply:
Also you can Change `ValidationRule` to this instead: 
```ts
type ValidationRule =
	| ValidationRuleObject
	| ValidationRuleObject[]
	| ValidationRuleName
	| { [ key: string ]: any }
```

Reply:
That error is because the `ValidationRule` has not been sufficiently type narrowed.  Since `schema` can be a `ValidationRuleName` or `ValidationRuleObject[]`, you can't be guaranteed access to the property.  Adding the index signature to `type ValidationRule` doesn't change that -- the type still has to be narrowed via a typeguard in order to properly access the object property.  Something like this works:
```ts
import Validator, { CheckerFunctionV2 } from 'fastest-validator';

const v = new Validator({ useNewCustomCheckerFunction: true })

const custom: CheckerFunctionV2 = function (value, errors, schema) {
  if (typeof schema === 'string') {
    console.log("custom --> schema", schema)
  } else if (Array.isArray(schema)) {
    schema.forEach((s, i) => {
      console.log(`custom --> schema[${i}]`, s.foo);
    });
  } else {
    console.log("custom -> schema.foo", schema.foo);
  }

  return value;
}

const check = v.compile({
  a: {
    type: "string",
    foo: "bar",
    custom,
  }
});
```

If the type of the `ruleSchema` can in fact be any of those types (hence why I've been asking about that), then the `custom` logic really needs to be set up to deal with the potential of getting any of the variants.

As I dig in a bit more here, I am starting to think that the type here shouldn't be `ValidationRule` but should instead of `ValidationRuleObject`.  It appears to be that the `rule.schema` passed to the custom validator always gets converted to an object and neither the string variant nor the array variant will ever get passed to the custom checker function.  If I make that change, then `foo` should be accessible on the `schema` without the need for typeguards.  Please let me know if you agree that what gets passed here will always be the object.  If so, the consumer can even do something like this, and get type safety in the custom properties:

```ts
type CustomRule = RuleCustom & { foo?: string };

const custom: CheckerFunctionV2 = function (value, errors, schema: CustomRule) {
  console.log("custom -> schema.foo", schema.foo); // string | undefined

  return value;
}
```